Pygrub extensions for XenServer:
 - 'not really' optionwhich just printso out what we would have done.
 - always boot 'xenkernel' and 'xeninitrd' if they exist

diff -r c88ae6cb060c tools/pygrub/src/pygrub
--- a/tools/pygrub/src/pygrub
+++ b/tools/pygrub/src/pygrub
@@ -761,6 +761,27 @@ if __name__ == "__main__":
     # get list of offsets into file which start partitions
     part_offs = get_partition_offsets(file)
 
+    if len(part_offs) < 1:
+        raise RuntimeError, "Disk has no partitions"
+
+    fs = fsimage.open(file, part_offs[0], bootfsoptions)
+
+    # We always boot the "default" kernel if it exists, rather than
+    # parsing the grub menu
+    try:
+        if fs.file_exists("/xenkernel"):
+            incfg["kernel"] = "/xenkernel"
+            incfg["args"] = default_args
+            if fs.file_exists("/xeninitrd"):
+                incfg["ramdisk"] = "/xeninitrd"
+        elif fs.file_exists("/boot/xenkernel"):
+            incfg["kernel"] = "/boot/xenkernel"
+            incfg["args"] = default_args
+            if fs.file_exists("/boot/xeninitrd"):
+                incfg["ramdisk"] = "/boot/xeninitrd"
+    except:
+        raise RuntimeError, "Unable to find partition containing kernel"
+
     for offset in part_offs:
         try:
             fs = fsimage.open(file, offset, bootfsoptions)
